serial-settings/src/lib.rs

```diff
@@ -0,0 +1,418 @@
+#![no_std]
+
+use core::fmt::Write;
+use embedded_storage::nor_flash::NorFlash;
+use miniconf::JsonCoreSlash;
+use serde::{Deserialize, Serialize};
+
+pub use menu;
+
+pub trait Settings:
+    for<'a> JsonCoreSlash<'a>
+    + Serialize
+    + Clone
+    + 'static
+    + for<'a> Deserialize<'a>
+{
+    fn reset(&mut self);
+}
+
+pub trait Interface:
+    embedded_io::Read
+    + embedded_io::ReadReady
+    + embedded_io::Write
+    + embedded_io::WriteReady
+{
+}
+
+impl<T> Interface for T where
+    T: embedded_io::Read
+        + embedded_io::ReadReady
+        + embedded_io::Write
+        + embedded_io::WriteReady
+{
+}
+
+pub trait Platform {
+    fn reset(&mut self, _interface: impl embedded_io::Write) {}
+
+    fn dfu(&mut self, mut interface: impl embedded_io::Write) {
+        self.reset(&mut interface);
+        writeln!(
+            &mut interface,
+            "Reset to DFU is not supported on this device"
+        )
+        .ok();
+    }
+
+    fn service(&mut self, mut interface: impl embedded_io::Write) {
+        writeln!(&mut interface, "Service data not available").ok();
+    }
+}
+
+impl<T> Platform for T {}
+
+pub struct Context<
+    'a,
+    I: Interface,
+    S: Settings,
+    P: Platform,
+    Flash: NorFlash + 'static,
```

Comment:
This implies that the settings will be persisted to flash, but it's also perfectly valid to persist them to something else, like EEPROM. Ultimately, this just needs to be some non-volatile memory location.

Actually, thinking about it, it doesn't even have to be non-volatile. Should we look for a different interface to use for this?

Reply:
I think https://docs.rs/embedded-storage/latest/embedded_storage/trait.Storage.html is the correct trait bound here instead.

Reply:
What about moving `save()` into the `Settings` trait and remove it completely from `Context`?

Reply:
Or move it into `Platform`

Reply:
Also maybe move `Settings` and `Interface` into aassociated types of `Platform`?

Reply:
I moved the various types into associated types of `Platform`, but I haven't touched `save()`.

I'm trying to understand why we would want to move it. Right now, it's abstracting away the concept of saving settings from the user. If we made it an associated function of `Platform`, that means the user could override it with some custom saving method. However, how does that make sense if we're internally loading the settings from the provided `storage` mechanism during initialization? If they were to save it differently, we couldn't load those settings.

It seems like what's currently implemented should be fine for now.